File: workers/go/spec_tables.go
package main

// spec_tables.go loads the shared case/fold tables that docs/WORKERS.md section 2 names as the
// rule: workers/spec/latin-lower.json and workers/spec/latin-fold.json.
//
// The tables are read at RUN TIME from workers/spec/ (see README.md -> "Strategy"). The strings of
// the spec files are ASCII in both files (latin-fold.json is all ASCII; latin-lower.json has no
// string values at all), so decoding them with encoding/json cannot introduce a chcp/GBK problem.
//
// stdlib only: encoding/json, fmt, os, path/filepath, runtime, strings.

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

// lowerEntry: target is always a single code point (the contract: "one code point to one code point").
type lowerEntry struct {
	target rune
}

// foldEntry: target is an ASCII string of one or two bytes (the contract: "an ASCII string of one
// or two characters"). Kept as a []byte so appending it to the output cannot re-encode anything.
type foldEntry struct {
	target []byte
}

// textTables is the whole shared-table state. Nothing else in the program consults a Unicode table:
// the Go standard library is not asked to lowercase, fold or normalize anything.
//
// The tables are dense slices indexed by code point, not maps: a map would be fine for correctness
// but a slice keeps the hot loop branch-free and, more importantly, keeps map iteration out of the
// program entirely. The contract's tables cover U+0000-U+024F (the shipped files top out at U+0236
// and U+024E keys), and a slice grown to the largest key in the file also survives a regenerated
// table that adds a key the contract's domain note did not mention.
type textTables struct {
	lower []*lowerEntry
	fold  []*foldEntry
}

var tables *textTables

type specTableFile struct {
	Map map[string]json.RawMessage `json:"map"`
}

// LoadTables reads both spec files. The directory is resolved by searchRepoSpec():
// next to the executable first, then walking up from the working directory.
func LoadTables() (*textTables, error) {
	dir, err := searchRepoSpec()
	if err != nil {
		return nil, err
	}
	lower, err := loadLowerTable(filepath.Join(dir, "latin-lower.json"))
	if err != nil {
		return nil, err
	}
	fold, err := loadFoldTable(filepath.Join(dir, "latin-fold.json"))
	if err != nil {
		return nil, err
	}
	t := &textTables{}
	if err := t.fill(lower, fold); err != nil {
		return nil, err
	}
	return t, nil
}

func loadLowerTable(path string) ([]struct {
	from int
	to   int
}, error) {
	raw, err := readSpecFile(path)
	if err != nil {
		return nil, err
	}
	var file specTableFile
	if err := json.Unmarshal(raw, &file); err != nil {
		return nil, fmt.Errorf("%s: not valid JSON: %v", path, err)
	}
	if file.Map == nil {
		return nil, fmt.Errorf("%s: no \"map\" object", path)
	}
	// Sorted output: Go map iteration order is randomised, and a table built in a random order would
	// not be observable in the output, but the duplicate check below must report a stable first
	// offender, so keys are sorted explicitly rather than iterated.
	keys := sortedJSONKeys(file.Map)
	out := make([]struct {
		from int
		to   int
	}, 0, len(keys))
	for _, key := range keys {
		from, err := parseDecimalKey(path, key)
		if err != nil {
			return nil, err
		}
		var to int
		if err := json.Unmarshal(file.Map[key], &to); err != nil {
			return nil, fmt.Errorf("%s: value for key %q is not an integer", path, key)
		}
		out = append(out, struct {
			from int
			to   int
		}{from, to})
	}
	return out, nil
}

func loadFoldTable(path string) ([]struct {
	from int
	to   string
}, error) {
	raw, err := readSpecFile(path)
	if err != nil {
		return nil, err
	}
	var file specTableFile
	if err := json.Unmarshal(raw, &file); err != nil {
		return nil, fmt.Errorf("%s: not valid JSON: %v", path, err)
	}
	if file.Map == nil {
		return nil, fmt.Errorf("%s: no \"map\" object", path)
	}
	keys := sortedJSONKeys(file.Map)
	out := make([]struct {
		from int
		to   string
	}, 0, len(keys))
	for _, key := range keys {
		from, err := parseDecimalKey(path, key)
		if err != nil {
			return nil, err
		}
		var to string
		if err := json.Unmarshal(file.Map[key], &to); err != nil {
			return nil, fmt.Errorf("%s: value for key %q is not a string", path, key)
		}
		out = append(out, struct {
			from int
			to   string
		}{from, to})
	}
	return out, nil
}

func readSpecFile(path string) ([]byte, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("cannot read the shared table %s: %v", path, err)
	}
	if len(raw) >= 3 && raw[0] == 0xEF && raw[1] == 0xBB && raw[2] == 0xBF {
		raw = raw[3:] // tolerate a BOM
	}
	return raw, nil
}

func parseDecimalKey(path, key string) (int, error) {
	if key == "" {
		return 0, fmt.Errorf("%s: empty table key", path)
	}
	value := 0
	for i := 0; i < len(key); i++ {
		if key[i] < '0' || key[i] > '9' {
			return 0, fmt.Errorf("%s: table key %q is not a decimal code point", path, key)
		}
		value = value*10 + int(key[i]-'0')
		if value > 0x10FFFF {
			return 0, fmt.Errorf("%s: table key %q is out of range", path, key)
		}
	}
	return value, nil
}

// sortedJSONKeys returns the keys of the decoded "map" object in ascending order, without letting
// an unsorted map iteration escape (the determinism rule of the contract).
func sortedJSONKeys(m map[string]json.RawMessage) []string {
	keys := make([]string, 0, len(m))
	for key := range m {
		keys = append(keys, key)
	}
	for i := 1; i < len(keys); i++ {
		for j := i; j > 0 && keys[j] < keys[j-1]; j-- {
			keys[j], keys[j-1] = keys[j-1], keys[j]
		}
	}
	return keys
}

// fill copies the two spec tables into slices indexed by code point. The contract's table domain is
// U+0000-U+024F; a key outside it is refused rather than silently truncated, because a table the
// contract does not describe would change the normalize output without anyone noticing.
func (t *textTables) fill(lower []struct {
	from int
	to   int
}, fold []struct {
	from int
	to   string
}) error {
	const domainMax = 0x024F

	lowerMax := 0
	for _, entry := range lower {
		if entry.from < 0 || entry.from > domainMax {
			return fmt.Errorf("latin-lower.json: key %d is outside the contract domain U+0000-U+024F", entry.from)
		}
		if entry.from == 0x0130 {
			return fmt.Errorf("latin-lower.json: U+0130 is present, but the contract says it is deliberately absent")
		}
		if entry.from > lowerMax {
			lowerMax = entry.from
		}
	}
	t.lower = make([]*lowerEntry, lowerMax+1)
	for _, entry := range lower {
		if t.lower[entry.from] != nil {
			return fmt.Errorf("latin-lower.json: duplicate key %d", entry.from)
		}
		t.lower[entry.from] = &lowerEntry{target: rune(entry.to)}
	}

	foldMax := 0
	for _, entry := range fold {
		if entry.from < 0 || entry.from > domainMax {
			return fmt.Errorf("latin-fold.json: key %d is outside the contract domain U+0000-U+024F", entry.from)
		}
		if entry.to == "" {
			return fmt.Errorf("latin-fold.json: key %d folds to the empty string", entry.from)
		}
		for i := 0; i < len(entry.to); i++ {
			if entry.to[i] > 0x7F {
				return fmt.Errorf("latin-fold.json: key %d folds to a non-ASCII string", entry.from)
			}
		}
		if entry.from > foldMax {
			foldMax = entry.from
		}
	}
	t.fold = make([]*foldEntry, foldMax+1)
	for _, entry := range fold {
		if t.fold[entry.from] != nil {
			return fmt.Errorf("latin-fold.json: duplicate key %d", entry.from)
		}
		t.fold[entry.from] = &foldEntry{target: []byte(entry.to)}
	}
	return nil
}

// searchRepoSpec returns the directory holding the shared spec tables.
//
// Order: <exe dir>/spec, then <exe dir>/../../spec (= workers/go/dist/../.. = workers/spec), then
// every ancestor of the working directory looking for workers/spec. The walk covers both the
// contract's launch form ("workers/go/dist/vmltext.exe --capability <name>", run from the
// repository root) and a direct run from inside workers/go.
func searchRepoSpec() (string, error) {
	var tried []string
	consider := func(dir string) string {
		if dir == "" {
			return ""
		}
		cleaned := filepath.Clean(dir)
		candidate := filepath.Join(cleaned, "spec")
		if hasBothTables(candidate) {
			return candidate
		}
		tried = append(tried, candidate)
		return ""
	}

	if exe, err := os.Executable(); err == nil {
		exeDir := filepath.Dir(exe)
		if found := consider(exeDir); found != "" {
			return found, nil
		}
		if found := consider(filepath.Join(exeDir, "..", "..")); found != "" {
			return found, nil
		}
	}
	if cwd, err := os.Getwd(); err == nil {
		for dir := cwd; ; {
			if found := consider(dir); found != "" {
				return found, nil
			}
			parent := filepath.Dir(dir)
			if parent == dir {
				break
			}
			dir = parent
		}
	}
	return "", fmt.Errorf("cannot find the shared tables (latin-lower.json, latin-fold.json); tried: %s",
		strings.Join(tried, ", "))
}

func hasBothTables(dir string) bool {
	for _, name := range []string{"latin-lower.json", "latin-fold.json"} {
		info, err := os.Stat(filepath.Join(dir, name))
		if err != nil || info.IsDir() {
			return false
		}
	}
	return true
}

func goRuntimeVersion() string {
	return runtime.Version()
}

File: workers/csharp/src/SpecTables.cs
// SpecTables.cs - the two shared tables of docs/WORKERS.md section 2.
//
// "The case and fold tables are shared data, not per-language library behaviour: every
// implementation reads them (or embeds them at build time); nobody consults their own runtime's
// tables. A language whose standard library would do more is expected to do exactly what the tables
// say and no more."
//
// .NET is emphatically a language whose standard library would do more: char.ToLowerInvariant,
// string.ToLower, StringComparer and CultureInfo all ship Unicode case behaviour, and every one of
// them is *close enough to look right* on the corpus' easy cases. This worker therefore never calls
// any of them: the file below is the only source of case and fold data in the program, exactly as
// the contract requires. (The one place this file does use char.IsSurrogate is classification, not
// case mapping - a surrogate pair has to be recognised as one code point before any table can be
// consulted for it, and the table simply has no entry for an astral code point.)
//
// Loaded at run time from workers/spec/*.json rather than embedded at build time: the file is
// already UTF-8 JSON with no non-ASCII string values in either table, System.Text.Json is part of
// the BCL, and reading it at startup means the artifact records the working tree instead of the
// revision that built it. A missing or malformed table is a startup error (exit 2), never a silent
// fallback to the runtime's own tables - a worker that answered with .NET's Unicode data would look
// right on almost every case and be wrong on exactly the ones this layer exists to find.

using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Text.Json;

namespace Vml;

internal sealed class SpecTables
{
    /// <summary>Code point -> lowercase code point, or -1. One code point to one code point.</summary>
    private readonly int[] _lower;

    /// <summary>Code point -> ASCII fold string, or null. One or two characters.</summary>
    private readonly string?[] _fold;

    private SpecTables(int[] lower, string?[] fold)
    {
        _lower = lower;
        _fold = fold;
    }

    public int LowerCount { get; private init; }

    public int FoldCount { get; private init; }

    /// <summary>The path the tables were read from, for diagnostics and describe.</summary>
    public string Directory { get; private init; } = string.Empty;

    /// <summary>The contract's table domain: U+0000-U+024F.</summary>
    private const int DomainMax = 0x024F;

    /// <summary>U+0130 LATIN CAPITAL LETTER I WITH DOT ABOVE: deliberately absent from the table.</summary>
    private const int DotlessDecision = 0x0130;

    public static SpecTables Load(string directory)
    {
        string lowerPath = Path.Combine(directory, "latin-lower.json");
        string foldPath = Path.Combine(directory, "latin-fold.json");

        List<KeyValuePair<int, int>> lower = LoadIntegerTable(lowerPath);
        List<KeyValuePair<int, string>> fold = LoadStringTable(foldPath);

        int lowerMax = 0;
        foreach (KeyValuePair<int, int> entry in lower)
        {
            if (entry.Key > DomainMax)
            {
                throw new SpecTableException($"latin-lower.json: key {entry.Key} is outside the contract domain U+0000-U+024F");
            }

            if (entry.Key == DotlessDecision)
            {
                throw new SpecTableException("latin-lower.json: U+0130 is present, but the contract says it is deliberately absent");
            }

            if (entry.Value < 0 || entry.Value > 0x10FFFF)
            {
                throw new SpecTableException($"latin-lower.json: key {entry.Key} maps to {entry.Value}, which is not a code point");
            }

            if (entry.Key > lowerMax)
            {
                lowerMax = entry.Key;
            }
        }

        int foldMax = 0;
        foreach (KeyValuePair<int, string> entry in fold)
        {
            if (entry.Key > DomainMax)
            {
                throw new SpecTableException($"latin-fold.json: key {entry.Key} is outside the contract domain U+0000-U+024F");
            }

            if (entry.Value.Length == 0)
            {
                throw new SpecTableException($"latin-fold.json: key {entry.Key} folds to the empty string");
            }

            foreach (char c in entry.Value)
            {
                if (c > 0x7F)
                {
                    throw new SpecTableException($"latin-fold.json: key {entry.Key} folds to a non-ASCII string");
                }
            }

            if (entry.Key > foldMax)
            {
                foldMax = entry.Key;
            }
        }

        int[] lowerArray = new int[lowerMax + 1];
        Array.Fill(lowerArray, -1);
        foreach (KeyValuePair<int, int> entry in lower)
        {
            if (lowerArray[entry.Key] != -1)
            {
                throw new SpecTableException($"latin-lower.json: duplicate key {entry.Key}");
            }

            lowerArray[entry.Key] = entry.Value;
        }

        string?[] foldArray = new string?[foldMax + 1];
        foreach (KeyValuePair<int, string> entry in fold)
        {
            if (foldArray[entry.Key] != null)
            {
                throw new SpecTableException($"latin-fold.json: duplicate key {entry.Key}");
            }

            foldArray[entry.Key] = entry.Value;
        }

        return new SpecTables(lowerArray, foldArray)
        {
            LowerCount = lower.Count,
            FoldCount = fold.Count,
            Directory = directory,
        };
    }

    /// <summary>
    /// Finds workers/spec by walking up from the working directory, so the worker runs both in the
    /// contract's launch form (from the repository root) and directly from workers/csharp/dist.
    /// </summary>
    public static string FindSpecDirectory()
    {
        string? overrideDirectory = Environment.GetEnvironmentVariable("VML_SPEC_DIR");
        if (!string.IsNullOrEmpty(overrideDirectory) && HasBothTables(overrideDirectory))
        {
            return Path.GetFullPath(overrideDirectory);
        }

        DirectoryInfo? directory = null;
        try
        {
            // Fully qualified on purpose: this class has a property called Directory, which shadows
            // the type name inside its own scope.
            directory = new DirectoryInfo(System.IO.Directory.GetCurrentDirectory());
        }
        catch (Exception error) when (error is IOException or UnauthorizedAccessException or ArgumentException)
        {
            directory = null;
        }

        var tried = new List<string>();
        while (directory != null)
        {
            string candidate = Path.Combine(directory.FullName, "workers", "spec");
            if (HasBothTables(candidate))
            {
                return candidate;
            }

            tried.Add(candidate);
            directory = directory.Parent;
        }

        // The executable's own location as a second guess: workers/csharp/dist/vmltext.exe sits two
        // levels below the worker directory, so ../../spec is workers/spec in an unmodified tree.
        string? exeDirectory = Path.GetDirectoryName(Environment.ProcessPath);
        if (!string.IsNullOrEmpty(exeDirectory))
        {
            string candidate = Path.GetFullPath(Path.Combine(exeDirectory, "..", "..", "spec"));
            if (HasBothTables(candidate))
            {
                return candidate;
            }

            tried.Add(candidate);
        }

        throw new SpecTableException(
            "cannot find the shared tables (latin-lower.json, latin-fold.json); tried: " + string.Join(", ", tried));
    }

    private static bool HasBothTables(string directory)
    {
        foreach (string name in new[] { "latin-lower.json", "latin-fold.json" })
        {
            string path = Path.Combine(directory, name);
            if (!File.Exists(path))
            {
                return false;
            }
        }

        return true;
    }

    /// <summary>Step 3: the lower table, one code point to one code point, or the input unchanged.</summary>
    public int Lower(int codePoint)
    {
        if (codePoint >= 0 && codePoint < _lower.Length)
        {
            int target = _lower[codePoint];
            if (target >= 0)
            {
                return target;
            }
        }

        return codePoint;
    }

    /// <summary>Step 4: the fold table, or null when the code point is not in it.</summary>
    public string? Fold(int codePoint)
    {
        if (codePoint >= 0 && codePoint < _fold.Length)
        {
            return _fold[codePoint];
        }

        return null;
    }

    private static List<KeyValuePair<int, int>> LoadIntegerTable(string path)
    {
        var result = new List<KeyValuePair<int, int>>();
        foreach (KeyValuePair<string, JsonElement> entry in LoadMap(path))
        {
            int codePoint = ParseKey(path, entry.Key);
            if (entry.Value.ValueKind != JsonValueKind.Number || !entry.Value.TryGetInt32(out int target))
            {
                throw new SpecTableException($"{Path.GetFileName(path)}: value for key {entry.Key} is not an integer");
            }

            result.Add(new KeyValuePair<int, int>(codePoint, target));
        }

        result.Sort((a, b) => a.Key.CompareTo(b.Key));
        return result;
    }

    private static List<KeyValuePair<int, string>> LoadStringTable(string path)
    {
        var result = new List<KeyValuePair<int, string>>();
        foreach (KeyValuePair<string, JsonElement> entry in LoadMap(path))
        {
            int codePoint = ParseKey(path, entry.Key);
            if (entry.Value.ValueKind != JsonValueKind.String)
            {
                throw new SpecTableException($"{Path.GetFileName(path)}: value for key {entry.Key} is not a string");
            }

            result.Add(new KeyValuePair<int, string>(codePoint, entry.Value.GetString() ?? string.Empty));
        }

        result.Sort((a, b) => a.Key.CompareTo(b.Key));
        return result;
    }

    private static List<KeyValuePair<string, JsonElement>> LoadMap(string path)
    {
        if (!File.Exists(path))
        {
            throw new SpecTableException($"cannot read the shared table {path}: it does not exist");
        }

        byte[] raw;
        try
        {
            raw = File.ReadAllBytes(path);
        }
        catch (Exception error) when (error is IOException or UnauthorizedAccessException)
        {
            throw new SpecTableException($"cannot read the shared table {path}: {error.Message}");
        }

        if (raw.Length >= 3 && raw[0] == 0xEF && raw[1] == 0xBB && raw[2] == 0xBF)
        {
            raw = raw[3..];
        }

        var result = new List<KeyValuePair<string, JsonElement>>();
        try
        {
            using var document = JsonDocument.Parse(raw);
            if (!document.RootElement.TryGetProperty("map", out JsonElement map) || map.ValueKind != JsonValueKind.Object)
            {
                throw new SpecTableException($"{Path.GetFileName(path)}: no \"map\" object");
            }

            foreach (JsonProperty property in map.EnumerateObject())
            {
                // EnumerateObject yields the document's order, which for these files is ascending.
                // The order does not reach the output - both callers sort - but nothing here ever
                // iterates a Dictionary, so a table read cannot become a table order.
                result.Add(new KeyValuePair<string, JsonElement>(property.Name, property.Value.Clone()));
            }
        }
        catch (JsonException error)
        {
            throw new SpecTableException($"{Path.GetFileName(path)}: not valid JSON: {error.Message}");
        }

        return result;
    }

    private static int ParseKey(string path, string key)
    {
        if (key.Length == 0)
        {
            throw new SpecTableException($"{Path.GetFileName(path)}: empty table key");
        }

        int value = 0;
        foreach (char c in key)
        {
            // Parsed by hand rather than with int.Parse: a table key is a decimal code point in
            // ASCII, and int.Parse would accept a leading sign, whitespace or a group separator
            // according to the current culture.
            if (c < '0' || c > '9')
            {
                throw new SpecTableException($"{Path.GetFileName(path)}: table key \"{key}\" is not a decimal code point");
            }

            value = (value * 10) + (c - '0');
            if (value > 0x10FFFF)
            {
                throw new SpecTableException($"{Path.GetFileName(path)}: table key \"{key}\" is out of range");
            }
        }

        return value;
    }
}

/// <summary>A startup failure that makes the worker unusable: reported on stderr, exit 2.</summary>
internal sealed class SpecTableException : Exception
{
    public SpecTableException(string message)
        : base(message)
    {
    }
}
